src/Stan/Analysis/Visitor.hs

```diff
@@ -0,0 +1,115 @@
+{- |
+Copyright: (c) 2020 Kowainik
+SPDX-License-Identifier: MPL-2.0
+Maintainer: Kowainik <xrom.xkov@gmail.com>
+
+HIE AST visitor for single-pass traversal.
+-}
+
+module Stan.Analysis.Visitor
+    ( VisitorState (..)
+    , getFinalObservations
+    , addObservation
+    , addObservations
+    , addFixity
+    , addOpDecl
+
+    , Visitor (..)
+    , visitAst
+    ) where
+
+import Relude.Extra.Lens (Lens', lens, over)
+
+import Stan.Ghc.Compat (RealSrcSpan)
+import Stan.Hie.Compat (HieAST (..), HieASTs (..), HieFile (..), TypeIndex)
+import Stan.Inspection (inspectionId)
+import Stan.Inspection.Style (stan0301)
+import Stan.Observation (Observation, Observations, mkObservation)
+
+import qualified Data.HashMap.Strict as HM
+import qualified Data.Map.Strict as Map
+import qualified Slist as S
+
+
+{- | State for the 'Visitor' object that stores all values during a
+single HIE AST traversal.
+-}
+data VisitorState = VisitorState
+    { visitorStateObservations :: !Observations
+
+      -- Operators for STAN-0301
+    , visitorStateFixities     :: !(HashMap Text ())
+    , visitorStateOpDecls      :: !(HashMap Text RealSrcSpan)
+    }
+
+-- | Initial empty state.
+initialVisitorState :: VisitorState
+initialVisitorState = VisitorState
+    { visitorStateObservations = mempty
+    , visitorStateFixities     = mempty
+    , visitorStateOpDecls      = mempty
+    }
+
+finaliseState :: HieFile -> VisitorState -> Observations
+finaliseState hie VisitorState{..} =
+    let operatorsWithoutFixity = HM.difference visitorStateOpDecls visitorStateFixities
```

Comment:
Worth leaving a comment about how we figure out the absence of fixity :hammer_and_wrench: